core/src/main/java/com/linecorp/armeria/client/endpoint/PropertiesFileWatcherRegistry.java

```diff
@@ -0,0 +1,223 @@
+            } catch (IOException e) {
+                throw new IllegalArgumentException("failed to watch file " + resourceUrl.getFile(), e);
+            }
+            startFutureIfPossible();
+        }
+    }
+
+    /**
+     * Stops watching a properties file corresponding to the {@code resourceUrl}.
+     * @param resourceUrl url to stop watching
+     */
+    void deregister(URL resourceUrl) {
+        final File file = new File(resourceUrl.getFile());
+        final Path parentPath = file.getParentFile().toPath();
+
+        synchronized (this) {
+            if (!ctxRegistry.containsKey(getRealPath(resourceUrl))) {
+                return;
+            }
+            final PropertiesFileWatcherContext context = ctxRegistry.remove(getRealPath(resourceUrl));
+            final boolean existsTargetFiles = ctxRegistry.keySet().stream().anyMatch(
+                    key -> key.startsWith(parentPath));
+            if (!existsTargetFiles) {
+                context.key.cancel();
+            }
+            stopFutureIfPossible();
+        }
+    }
+
+    private void startFutureIfPossible() {
+        if (!isRunning() && !ctxRegistry.isEmpty()) {
+            future = CompletableFuture.runAsync(new PropertiesFileWatcherRunnable(), executor);
+        }
+    }
+
+    private void stopFutureIfPossible() {
+        checkState(future != null, "tried to stop null future");
+        if (isRunning() && ctxRegistry.isEmpty()) {
+            future.cancel(true);
+        }
+    }
+
+    private static Path getRealPath(URL url) {
+        return getRealPath(new File(url.getFile()));
+    }
+
+    private static Path getRealPath(File file) {
+        return getRealPath(file.toPath());
+    }
+
+    private static Path getRealPath(Path path) {
+        try {
+            return path.toRealPath();
+        } catch (IOException e) {
+            throw new IllegalArgumentException("failed to locate file " + path, e);
+        }
+    }
+
+    /**
+     * Check if future for {@code WatchService} is running.
+     * @return true if future is running
+     */
+    @VisibleForTesting
+    boolean isRunning() {
+        return future != null && !future.isDone();
+    }
+
+    /**
+     * Close watch service, future, and clear registry.
+     * @throws Exception may be thrown if {@code WatchService} is already closed
+     */
+    @Override
+    public void close() throws Exception {
+        ctxRegistry.clear();
+        stopFutureIfPossible();
+        watchService.close();
+    }
+
+    private class PropertiesFileWatcherRunnable implements Runnable {
+        @Override
+        public void run() {
+            try {
+                WatchKey key;
+                while ((key = watchService.take()) != null) {
+                    for (WatchEvent<?> event : key.pollEvents()) {
+                        @SuppressWarnings("unchecked")
+                        final Path watchedPath = getRealPath(
```

Comment:
Think you should catch and log the exception here and `continue` - it's possible `getRealPath` will throw if the file was modified then immediately deleted. Fine to skip such an event.